### PR TITLE
add package to module by command

### PR DESCRIPTION
**Pull request by zedisdog**
now module can require package by command
```bash
php artisan module:require packageName moduleName
```
ps: This package is awesome.could I push this?I know the code seems....some.....stupid.:expressionless:

### CODE REVIEW

**Comment by nWidart** [comment]
Hi and thank you for the contribution!

Could you please squash your commits so it makes just one commit please?

File: .gitignore [hunk 1]
```diff
@@ -4,3 +4,4 @@ composer.lock
 /fixture/vendor
 build
 .php_cs.cache
+.idea
```
Comment (nWidart) [inline .gitignore]: Please remove this. This should go in your global gitignore file.

File: composer.json [hunk 2]
```diff
@@ -18,7 +18,8 @@
     }
   ],
   "require": {
-    "php": ">=7.1"
+    "php": ">=7.1",
+    "ext-json": "*"
```
Comment (nWidart) [inline composer.json]: Is this extension required by something we used?

File: src/Commands/RequireCommand.php [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+     * @var string
+     */
+    protected $name = 'module:require';
+
+    /**
+     * The console command description.
+     *
+     * @var string
+     */
+    protected $description = 'add dependencies for the specified module.';
```
Comment (nWidart) [inline src/Commands/RequireCommand.php]: Please rename this to:
`Add a dependency for the specified module`

File: src/Commands/RequireCommand.php [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    /**
+     * Execute the console command.
+     */
+    public function handle()
+    {
+        if (!$this->argument('packageName')) {
+            $this->error('packageName is required');
+            return;
+        }
+        $this->getModule();
```
Comment (nWidart) [inline src/Commands/RequireCommand.php]: Why is this here? The return of the method isn't used.

File: src/Commands/RequireCommand.php [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+    protected $signature = 'module:require {--dev} {packageName} {module?}';
+
+    /**
+     * Execute the console command.
+     */
+    public function handle()
+    {
+        if (!$this->argument('packageName')) {
+            $this->error('packageName is required');
```
Comment (nWidart) [inline src/Commands/RequireCommand.php]: Please rename the text to : `Argument packageName is required`

File: src/Commands/RequireCommand.php [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+            $requireKey,
+            $requireOtherKey,
+            $newPackage,
+            $newPackageVersion,
+            $additionalScripts
+        );
+        $this->putComposer($moduleComposerPath, $moduleComposer);
+    }
+
+    protected function putComposer($path, $data)
```
Comment (nWidart) [inline src/Commands/RequireCommand.php]: Could you rename this method to `writeComposerFile`?

File: src/Commands/RequireCommand.php [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+            $this->info('package install success');
+            $this->info('updating module composer....');
+            $this->updateModuleComposer();
+            $this->info('updated module composer');
+        } else {
+            $this->warn('package install failed');
+        }
+    }
+
+    protected function getComposer($path)
```
Comment (nWidart) [inline src/Commands/RequireCommand.php]: Could you rename this method to `getComposerFileContent`

Comment (nWidart) [reply src/Commands/RequireCommand.php]: Also why is this method protected?

File: src/Commands/RequireCommand.php [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+    /**
+     * @param Json $moduleComposer
+     * @param string $requireKey
+     * @param string $requireOtherKey
+     * @param string $newPackage
+     * @param string $newPackageVersion
+     * @param array $additionalScripts
+     * @return Json
+     */
+    private function setModuleComposer($moduleComposer, $requireKey, $requireOtherKey, $newPackage, $newPackageVersion, $additionalScripts)
```
Comment (nWidart) [inline src/Commands/RequireCommand.php]: Can you please rename this to `updateModuleComposerFile`

File: src/Commands/RequireCommand.php [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+    protected function putComposer($path, $data)
+    {
+        file_put_contents($path, json_encode($data, JSON_UNESCAPED_SLASHES | JSON_PRETTY_PRINT));
+    }
+
+    /**
+     * @param array $composer
+     * @return array
+     */
+    private function getAdditional($composer)
```
Comment (nWidart) [inline src/Commands/RequireCommand.php]: The method name is confusing to me. Get additional what? Can you be explicit here please?

File: src/Commands/RequireCommand.php [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+        $newScripts = $composer['scripts'];
+        return array_diff_key($newScripts, $this->oldScripts);
+    }
+
+    /**
+     * @param array $composer
+     * @param string $require_key
+     * @return array
+     */
+    private function getNewPackageInfo($composer, $require_key)
```
Comment (nWidart) [inline src/Commands/RequireCommand.php]: Please use `camelCase` as the parameter names.

**Comment by nWidart** [comment]
Thank you for your work on this matter, however I'll have to close this PR for now.

I should probably document the recommended ways to handle this inside projects, based on how composer works.

**Comment by zedisdog** [comment]
Hi,thanks for your time!I think there are some misunderstandings.

This command will first require package in project root folder,then copy package info from `composer.json in root folder` to `composer.json in the module folder`.

So the vendor folder won't be generated in module folder.The package which be added will only be placed in root vendor folder.

if you want it,I'm glad to submit a PR again with fixes which you mentioned above.:beers:

**Comment by zedisdog** [comment]
Hi,by the way,I'm in china now.So you know,you work I sleep,you sleep I work.Please forgive me for my offline.:smile:
